src/mlpack/tests/sdp_primal_dual_test.cpp

```diff
@@ -557,92 +558,141 @@ BOOST_AUTO_TEST_CASE(CorrelationCoeffToySdp)
-
-//      // A_ij(j, i) = -1.
-//      aRef(neighbors(j, i), i) = -1;
-
-//      // A_ij(j, j) = 1.
-//      aRef(neighbors(j, i), neighbors(j, i)) = 1;
-
-//      // The constraint b_ij is the distance between these two points.
-//      sdp.SparseB()[index] = distances(j, i);
-//    }
-//  }
-
-//  return sdp;
-// }
-
-// /**
-// * Maximum variance unfolding
-// *
-// * Test doesn't work, because the constraint matrices are not linearly
-// * independent.
-// */
-// BOOST_AUTO_TEST_CASE(SmallMvuSdp)
-// {
-//  const size_t n = 20;
-
-//  arma::mat origData(3, n);
-
-//  // sample n random points on 3-dim unit sphere
-//  GaussianDistribution gauss(3);
-//  for (size_t i = 0; i < n; i++)
-//  {
-//    // how european of them
-//    origData.col(i) = arma::normalise(gauss.Random());
-//  }
-
-//  auto sdp = ConstructMvuSDP(origData, 5);
-
-//  PrimalDualSolver<SDP<arma::sp_mat>> solver(sdp);
-//  arma::mat X, Z;
-//  arma::vec ysparse, ydense;
-//  const auto p = solver.Optimize(X, ysparse, ydense, Z);
-//  BOOST_REQUIRE(p.first);
-// }
+/**
+ * Maximum variance unfolding (MVU) SDP to learn the unrolled gram matrix. For
+ * the SDP formulation, see:
+ *
+ *   Unsupervised learning of image manifolds by semidefinite programming.
+ *   Kilian Weinberger and Lawrence Saul. CVPR 04.
+ *   http://repository.upenn.edu/cgi/viewcontent.cgi?article=1000&context=cis_papers
+ *
+ * Here is another paper which clarifies more specific SDP formulation of MVU problem.
+ *
+ *   Scalable semidefinite manifold learning
+ *   Nikolaos Vasiloglou, Alexander G. Gray, David V. Anderson.
+ *   2008 IEEE Workshop on Machine Learning for Signal Processing
+ *
+ * @param origData origDim x numPoints
+ * @param numNeighbors
+ */
+static inline SDP<arma::sp_mat> ConstructMvuSDP(const arma::mat& origData,
+                                              size_t numNeighbors)
+{
+  const size_t numPoints = origData.n_cols;
+
+  assert(numNeighbors <= numPoints);
+
+  arma::Mat<size_t> neighbors;
+  arma::mat distances;
+  KNN knn(origData);
+  knn.Search(numNeighbors, neighbors, distances);
+
+  // how to construct the matrices A, b, and C? Plz refer to the above paper.
+  std::set<std::pair<int, int>> neighbors_set;
+  for (size_t i = 0; i < neighbors.n_cols; ++i)
+  {
+    for (size_t j = 0; j < numNeighbors; ++j)
+    {
+      size_t x = i, y = neighbors(j, i);
+      if (x > y)
+      {
+        x = neighbors(j, i);
+        y = i;
+      }
+
+      if (neighbors_set.find(std::make_pair(x, y)) != neighbors_set.end())
+      {
+        continue;
+      }
+
+      neighbors_set.insert(std::make_pair(x, y));
+    }
+  }
+
+  // SDP<arma::sp_mat> sdp(numPoints, numNeighbors * numPoints, 1);
+  SDP<arma::sp_mat> sdp(numPoints, neighbors_set.size(), 1);
+  sdp.C().eye(numPoints, numPoints);
+  sdp.C() *= -1;
+  sdp.DenseA()[0].ones(numPoints, numPoints);
+  sdp.DenseB()[0] = 0;
+  neighbors_set.clear();
+  size_t _index = 0;
+
+  for (size_t i = 0; i < neighbors.n_cols; ++i)
```

Comment:
It seems like you could just loop over the already-filled `neighbors_set` here instead of refiltering all the duplicate neighbor pairs.

Reply:
right! I planned to loop the already-filled set, but then I found I cannot determine `sdp.SparseB()[index] = distances(j, i);` without `j`. Though we can store the `distances(j, i)` into the set, I don't think it is sensible to do it for the distances are floating-point numbers.

Reply:
You could use `std::tuple<size_t, size_t, double>` instead of `std::pair<int, int>`.  (Actually it should also be `size_t` instead of `int` regardless of what you do.)

Reply:
good suggestion! I will update it, but I concern about the comparison of floating-point number. I will figure it out.

Reply:
overloading the comparabel function will solve the concerning:)